base_layer/p2p/src/handlers.rs

```diff
@@ -0,0 +1,309 @@
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::tari_message::{BlockchainMessage, NetMessage, PeerMessage, TariMessageType, ValidatorNodeMessage};
+use tari_comms::{
+    dispatcher::{DispatchError, DispatchResolver},
+    message::{DomainMessageContext, MessageHeader},
+    types::DomainMessageDispatcher,
+};
+use tari_crypto::keys::PublicKey;
+
+// Create a common variable that the worker can change and the test can read to determine that the message was
+// correctly dispatched
+#[cfg(test)]
+static mut TEST_CALLED_FN_TYPE: u8 = 0;
+
+#[derive(Clone)]
+pub struct DomainInboundMessageResolver;
+
+impl<PubKey: PublicKey> DispatchResolver<TariMessageType, DomainMessageContext<PubKey>>
+    for DomainInboundMessageResolver
+{
+    /// The dispatch type is determined from the content of the DomainMessageContext, which is used to dispatch the
+    /// message to the correct domain handler
+    fn resolve(&self, message_context: &DomainMessageContext<PubKey>) -> Result<TariMessageType, DispatchError> {
+        let message_header: MessageHeader<TariMessageType> = message_context
+            .message
+            .to_header()
+            .map_err(|err| DispatchError::HandlerError(format!("{}", err)))?;
+        Ok(message_header.message_type)
+    }
+}
+
+/// Specify what handler function should be called for messages with different domain message types
+pub fn construct_domain_msg_dispatcher<PubKey: PublicKey>(
+) -> DomainMessageDispatcher<PubKey, TariMessageType, DomainInboundMessageResolver> {
+    DomainMessageDispatcher::<PubKey, TariMessageType, DomainInboundMessageResolver>::new(
+        DomainInboundMessageResolver {},
+    )
+    .route(TariMessageType::from(NetMessage::Join), handler_net_message_join)
```

Comment:
Still not convinced we need this TariMessageType wrapper - why can't we simply use `MessageHeader<u8>` and dispatch e.g. ` NetMessage::Join`? If we need to check for example `TariMessageType::is_known_message` (perhaps we might want to use this in the resolver?) we can wrap the u8 there, return a 'resolved' u8 and not we can keep the Dispatcher routing neater without `TariMessageType::from`s (or `NetMessage::xxxx.into()`)

Reply:
I prefer having it as a TariMessageType as this can be extend in the future, having it as a u8 does not link it to what it actually is. As an alternative we could change the `from`s  and rather use `into`s, it does make it look nicer.

Reply:
I am pretty sure that when you define the `from` you get the `into` for free just need to use it.

Reply:
Changed it to `into`s